ddd/ddd-05-mem-nonblocking.DISCUSSION.md

```diff
@@ -0,0 +1,250 @@
+      between send/recv and BIO_read/BIO_write calls; and
+
+    - the buffer passed to BIO_read() and recv() must always be big
+      enough to receive the largest possible packet, else silent
+      truncation occurs.
+
+    The latter is an unfortunate design issue with the Berkeley sockets API
+    for SOCK_DGRAM, which is replicated here in BIO_read which simply serves
+    as a wrapper for it. The caller thus has to know the worst-case UDP
+    datagram payload size, but knowing this requires knowledge of the
+    network or at least the configuration of local network interfaces.
+
+    (Random factoid: this design issue once in the past led me to determine
+    the maximum size of an incoming UDP datagram as
+
+      max(64 KiB, max(MTU of every local interface))
+
+    -- reason being that an IPv6 fragmentation header has a 16-bit length field,
+    so an IP packet greater than 64 KiB (using the IPv6 jumbogram header) can
+    only be sent if you have an interface with a native MTU exceeding it.
+    Another option is to use MSG_PEEK | MSG_TRUNC on systems which support it
+    but this requires calling recv twice per packet.)
+
+  - For the buffer size issue, our options for supporting existing applications
+    include:
+
+    - Using a maximum MTU of 1472 by default; then any application using
+      old BIO_read/BIO_write calls will work so long as it uses a buffer
+      of at least 1472 bytes (and there is a 1:1 correspondance of BIO calls and
+      syscalls).
+
+    - Applications which can support larger MTUs (e.g. a network with
+      jumbo frames) could indicate support via some method TBD, such as
+      capability flags (see below).
+
+    In this model, for the purposes of compatibility and minimising changes:
+
+    - BIO_ctrl_get_write_guarantee (BIO_C_GET_WRITE_GUARANTEE) would always
+      return either 0 or a value >=1472. (This is a change from BIO_dgram,
+      which does not implement BIO_C_GET_WRITE_GUARANTEE.)
+
+    - BIO_ctrl_pending (BIO_CTRL_PENDING) returns the size of the next frame
+      which will be popped by BIO_read. It's not possible to determine how many
+      frames are queued in a frame-BIO or the total amount of data queued by
+      using this function. (This is a change from BIO_dgram, which always
+      returns 0 for BIO_CTRL_PENDING.)
+
+    The following APIs could be provided for applications which want to
+    support larger MTUs (new applications, and applications willing to make
+    larger changes):
+
+    ```c
+      /*
+       * A flag settable on BIO_dgram_mem. A BIO_read call will not truncate,
+       * but instead fail if the frame is too large for the provided buffer. The
+       * application must call BIO_ctrl_pending() to determine the necessary
+       * buffer size and try again.
+       */
+      int BIO_dgram_set_no_trunc(bio, 1);
+
+      /*
+       * Returns libssl's idea of what our maximum UDP payload size is.
```

Comment:
This is odd....asking a BIO about what libssl's idea of something is. Libssl should surely ask the BIO what the MTU is and use that not the other way around.

Reply:
For `BIO_dgram` we have `BIO_CTRL_DGRAM_GET/SET_MTU` which `libssl` can make use of to find an MTU communicated by the application. So I guess the application would create the `BIO_dgram_mem`, which defaults to an MTU (more accurately, maximum UDP payload length) of 1472, and can then set a larger MTU using `BIO_CTRL_DGRAM_SET_MTU` if it wishes.

IIRC, I believe where I was going with `BIO_get_max_dgram_len` is that we should have a function that more capable applications can use to ask the QUIC stack what maximum UDP payload length it would *like* to be able to use. A better name for this would probably be `SSL_get_preferred_max_dgram_len`.

So basically, for a basic client:
- `char buf[1472 (or larger)];`
- never calls `BIO_CTRL_DGRAM_SET_MTU`, `BIO_dgram_mem` defaults to 1472
- libssl calls `BIO_CTRL_DGRAM_GET_MTU` and determines it can use a UDP payload size of 1472

For a better client:
```
ssize_t maxlen = SSL_get_preferred_max_dgram_len(ssl);
...
char *buf = malloc(maxlen);
BIO_CTRL_DGRAM_SET_MTU(maxlen);
```

Reply:
Ah, right so this is really an SSL function. I find the BIO_* prefix used here as confusing.